File: src/main/scala/category/design/Trie.scala
package category.design

/**
  * 208. Implement Trie (Prefix Tree)
  */
class Trie() {
  
  /** Initialize your data structure here. */
  var word    : String            = _
  var children: Array[Trie] = new Array[Trie](26)
  
  /** Inserts a word into the trie. */
  def insert(word: String): Unit = {
    var curr = this
    word.foreach { c =>
      if (curr.children(c - 'a') == null) curr.children(c - 'a') = new Trie
      curr = curr.children(c - 'a')
    }
    curr.word = word
  }
  
  /** Returns if the word is in the trie. */
  def search(word: String): Boolean = {
    var curr = this
    word.foreach { c =>
      curr = curr.children(c - 'a')
      if (curr == null) return false
    }
    curr.word != null
  }
  
  /** Returns if there is any word in the trie that starts with the given prefix. */
  def startsWith(prefix: String): Boolean = {
    var curr = this
    prefix.foreach { c =>
      curr = curr.children(c - 'a')
      if (curr == null) return false
    }
    curr != null
  }
}

File: src/main/python/Trie.py
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.lookup = {}

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        current = self.lookup
        for c in word:
            if c not in current: current[c] = {}
            current = current[c]
        current["_"] = True

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        current = self.lookup
        for c in word:
            if c not in current: return False
            current = current[c]
        return "_" in current

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        current = self.lookup
        for c in prefix:
            if c not in current: return False
            current = current[c]
        return True


if __name__ == '__main__':
    instance = Trie()
    instance.insert("apple")
    assert instance.search("apple")
    assert not instance.search("app")
    assert instance.startsWith("app")
    instance.insert("app")
    assert instance.search("app")
